### PR TITLE
Create a UVM agent for ISA coverage

### PR DESCRIPTION
**Pull request by silabs-robin**
This fulfils the initial criteria of https://github.com/openhwgroup/core-v-verif/issues/488

It introduces an agent which monitors activity in the 40x's tracer module (rvfi should be used later https://github.com/openhwgroup/core-v-verif/issues/489).
The monitor decodes/disassembles the received instruction and parses it into a uvm_object representing the instruction (later, https://github.com/openhwgroup/core-v-verif/issues/487 should enable full support for decoding).
The uvm object is passed to the coverage model which samples into covergroups defined for RV32IMCZifencei_Zicsr (more tweaks and refinements later https://github.com/openhwgroup/core-v-verif/issues/490 and bottom of https://github.com/openhwgroup/core-v-verif/issues/488).

The name "uvma_isa" was chosen (instead of the prescribed "uvma_covg" or "uvma_isa_covg") because it matches the existing naming-patterns for both 1) other agents, and 2) covergroups.

More polish is needed (as noted above), but the system stands by itself and demonstrates a complete architecture and coverage is disabled by default.

### CODE REVIEW

**Comment by silabs-robin** [comment]
Usage:
1. Within `cv32e40x/env/uvme/uvme_cv32e40x_cfg.sv`, set `isa_cfg.enabled` to 1 (is there a better way to toggle this?).
2. Run any test in uvmt with `COV=1`.
3. Use your preferred simulator and coverage database viewer.

File: cv32e40x/env/uvme/uvme_cv32e40x_cfg.sv [hunk 1]
```diff
@@ -35,6 +35,7 @@ class uvme_cv32e40x_cfg_c extends uvm_object;
    //rand int unsigned             debug_clk_period;
    
    // Agent cfg handles
+   rand uvma_isa_cfg_c        isa_cfg;
```
Comment (MikeOpenHWGroup) [inline cv32e40x/env/uvme/uvme_cv32e40x_cfg.sv:38]: Its not clear from the class or object name that this is a configuration object for a functional coverage model.  Perhaps `uvma_isa_fcov_cfg_c`?

File: lib/uvm_agents/uvma_isa/uvma_isa_mon.sv [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  fork
+    begin
+      forever begin
+        uvma_isa_mon_trn_c mon_trn;
+
+        @(cntxt.vif.retire);
+
+        mon_trn = new();
+        mon_trn.instr = new();
+        mon_trn.instr.name =
```
Comment (MikeOpenHWGroup) [inline lib/uvm_agents/uvma_isa/uvma_isa_mon.sv:66]: This is hard to follow.   Either:
- use a case statement (nested, if required), or
- improve the formatting for readability.

File: lib/uvm_agents/uvma_isa/uvma_isa_if.sv [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+//     https://solderpad.org/licenses/
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+interface uvma_isa_if;
```
Comment (MikeOpenHWGroup) [inline lib/uvm_agents/uvma_isa/uvma_isa_if.sv:17]: Where is this instantiated in the environment?   Is another component expected to source 32-bit instructions on this interface?

**Review by MikeOpenHWGroup — CHANGES_REQUESTED** [review]
Thanks for this @silabs-robin.  I am _very_ pleased to see this work starting - this coverage model is hugely important for all CORE-V cores and it will be heavily re-used.  I have added a few comments, most of them minor and there are two that are more important than the others:
1. Please give some thought to a more descriptive name for this agent.  "isa" is very generic and does not convey that this is a coverage model.
2. I found the code in `uvma_isa_mon_c::run_phase()` very difficult to follow (even though I am familiar with the ISA).  Se my comments in that file.

File: cv32e40x/env/uvme/uvme_cv32e40x_cfg.sv [hunk 4]
```diff
@@ -90,8 +92,15 @@ class uvme_cv32e40x_cfg_c extends uvm_object;
       obi_instr_cfg.read_enabled  == 1;
       obi_data_cfg.write_enabled  == 1;
       obi_data_cfg.read_enabled   == 1;
+      isa_cfg.enabled              == 0;
```
[the remark this answers is not in the document]
Comment (strichmo) [reply cv32e40x/env/uvme/uvme_cv32e40x_cfg.sv:95]: @MikeOpenHWGroup This was my idea.  This is a very early version of the new func cov model and really we shouldn't be using it until it's more functional (at least with IMC coverage).  Therefore to avoid polluting fcov stats in any regressions I proposed to leave this off and contributors who work to update the model will enable it via a local edit.  Should add a FIXME here however at the least.

File: cv32e40x/env/uvme/uvme_cv32e40x_pkg.sv [hunk 5]
```diff
@@ -38,6 +38,7 @@ package uvme_cv32e40x_pkg;
    import uvml_hrtbt_pkg  ::*;
    import uvml_sb_pkg     ::*;
    import uvml_trn_pkg    ::*;  
+   import uvma_isa_pkg::*;
```
[the remark this answers is not in the document]
Comment (strichmo) [reply cv32e40x/env/uvme/uvme_cv32e40x_pkg.sv:41]: Agreed, this agent's main purpose is to take ISA-based functional coverage so cov has to end up in the agent name.  ISA agent is just too general.

File: lib/uvm_agents/uvma_isa/instr.sv [hunk 6]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 OpenHW Group
```
[the remark this answers is not in the document]
Comment (strichmo) [reply lib/uvm_agents/uvma_isa/instr.sv:1]: This should be named in accordance with the agent and should include a trn suffix.  Since this will always be a passive agent, it's not a sequence item, but certainly is a "transaction" within the framework/terminology of our OpenHW agents.

File: lib/uvm_agents/uvma_isa/uvma_isa_if.sv (hunk 3 above)
Comment (strichmo) [reply lib/uvm_agents/uvma_isa/uvma_isa_if.sv:17]: Perhaps add a little bit more self-documentation here, that this is a temporary interface.  I would imagine that we would replace this with an exposed analysis port for OBI (or RVFI directly, if ready).

File: lib/uvm_agents/uvma_isa/uvma_isa_mon.sv (hunk 2 above)
Comment (strichmo) [reply lib/uvm_agents/uvma_isa/uvma_isa_mon.sv:66]: Note that this will be replaced with another decoder solution.

Comment (MikeOpenHWGroup) [reply lib/uvm_agents/uvma_isa/uvma_isa_mon.sv:66]: OK

**Comment by silabs-robin** [comment]
Thanks you @MikeOpenHWGroup  and @strichmo  for reviewing this PR.
I agree with all of the feedback and have tried to address it in the latest commits.
(I'll reply to some of the comments above with some elaboration too).

File: lib/uvm_agents/uvma_isa/instr.sv (hunk 6 above)
[the remark this answers is not in the document]
Comment (silabs-robin) [reply lib/uvm_agents/uvma_isa/instr.sv]: I don't see a solution that satisfies all constraints.

If one leave it as is, it doesn't follow the repo's naming scheme.
If the name reflects "rv32imcz", it conflicts with the task's "ability to extend beyond".
If suffixing "_trn", it should extend uvml_trn_mon_trn_c and conflicts with mon_trn.
If mon_trn holds the data instead, then the name reflects "instruction" even less.
If "instr" is a struct typedef, then it conflicts with "standard UVM object conventions to model instructions".

For now, I've changed it to "uvma_isacov_instr".

File: cv32e40x/env/uvme/uvme_cv32e40x_pkg.sv (hunk 5 above)
[the remark this answers is not in the document]
Comment (silabs-robin) [reply cv32e40x/env/uvme/uvme_cv32e40x_pkg.sv]: I changed the name to "uvma_isacov", but if we would like something else I can set it to that.

Rationale:
If possible, I would prefer to keep the convention of "uvma_<name>" and avoid "uvma_<name1>_<name2>", because the former is consistent and simple while the latter adds a cognitive overhead in deciphering each term of our multi-word names (e.g. "uvma_abc_xyz_mon_trn_logger").
The reason I avoided "uvma_isa**f**cov" is 1) too many codes and abbreviations are difficult, and 2a) "isa coverage" names a type of coverage and 2b) arguably suggests a sub type of "functional coverage" (isa declares functionality, is not rtl, is not e.g. performance reqs).
The only precedence of "fcov" seems to be in "isa_fcov_holes", and all examples of "covg" seem to be for classes containing covergroups.

File: lib/uvm_agents/uvma_isa/uvma_isa_if.sv (hunk 3 above)
Comment (silabs-robin) [reply lib/uvm_agents/uvma_isa/uvma_isa_if.sv]: I might have misunderstood "Input of coverage model comes from an analysis port"?
This way of using interface is the same as all the other "Agent interfaces" in the tb.
But maybe the task's intention was to use TLM ports for input to the agent itself, and not just to the "cov_model" as it is now.

I added a TODO comment about updating the interface, both regarding RVFI and potentially TLM port.

**Review by MikeOpenHWGroup — APPROVED** [review]

**Comment by MikeOpenHWGroup** [comment]
Thanks @silabs-robin, the updates look good.  Will merge this in.
